File: algorithm/55.c
#include <stdio.h>

int canJump(int *nums, int numsSize)
{
    // the longest point we can get to.
    int max = nums[0];
    for (int i = 1; i < numsSize; i++)
    {
        if (i > max)
        {
            return 0;
        }
        if (i + nums[i] > max)
        {
            max = i + nums[i];
        }
    }
    return 1;
}

int main(int argc, char *argv[])
{
    int a[5] = {2, 3, 1, 1, 4};
    printf("%d\n", canJump(a, 5));
    return 0;
}

File: go/goalgor/algor/55.go
func canJump(nums []int) bool {
	n := len(nums)
	rightmost := 0
	for i := 0; i < n; i++ {
		if i <= rightmost {
			if rightmost < i+nums[i] {
				rightmost = i + nums[i]
			}
			if rightmost >= n-1 {
				return true
			}
		}

	}
	return false
}

// Add dp
// func canJump(nums []int) bool {
//     n := len(nums)
//     can := make([]int, n)
//     can[0]=1
//     for i:=0;i<n;i++ {
//         for j:=i-1;j>=0;j-- {
//             if can[j]==1 && (i-j) <= nums[j] {
//                 can[i]=1
//                 break
//             }
//         }
//     }
//     return can[n-1] == 1
// }